packages/webservice/app/src/main/kotlin/com/yahoo/navi/ws/utilities/CarbonQueryLogger.kt

```diff
@@ -0,0 +1,156 @@
+/**
+ * Copyright 2020, Yahoo Holdings Inc.
+ * Licensed under the terms of the MIT license. See accompanying LICENSE.md file for terms.
+ */
+package com.yahoo.navi.ws.utilities
+
+import com.google.common.collect.Iterables
+import com.yahoo.elide.datastores.aggregation.core.QueryLogger
+import com.yahoo.elide.datastores.aggregation.core.QueryResponse
+import com.yahoo.elide.datastores.aggregation.query.Query
+import com.yahoo.elide.security.User
+import com.yahoo.navi.ws.models.beans.utilities.QueryStats
+import com.yahoo.navi.ws.models.beans.utilities.QueryStatus
+import java.net.InetAddress.getLocalHost
+import java.util.Optional
+import java.util.Timer
+import java.util.TimerTask
+import java.util.UUID
+import javax.persistence.EntityManager
+import javax.persistence.PersistenceContext
+import javax.transaction.Transactional
+import javax.ws.rs.core.MultivaluedMap
+
+object Constants {
+    const val MAX_QUERIES = 5
+    const val TIMEOUT = 10000
+}
+
+@Transactional
+open class CarbonQueryLogger constructor(
+    maxQueries: Int = Constants.MAX_QUERIES,
+    timeOut: Long = Constants.TIMEOUT.toLong()
+) : QueryLogger, TimerTask() {
+
+    private var queriesToBeLogged: MutableList<QueryStats> = ArrayList()
+    private var queryWatcher: MutableMap<UUID, QueryStats> = HashMap()
+
+    @PersistenceContext
+    private lateinit var entityManager: EntityManager
+
+    private var t: Timer = Timer()
+    private val maxQueries: Int = maxQueries
+    private val timeOut: Long = timeOut
+
+    init {
+        t.schedule(this, 0, this.timeOut)
+    }
+
+    override fun cancelQuery(queryId: UUID) {
+        var endTime: Long = System.currentTimeMillis()
+        var qstat: QueryStats = queryWatcher[queryId]!!
+        synchronized(this) {
+            qstat.status = QueryStatus.CANCELLED
+            qstat.durationInMillis = endTime - qstat.startTime!!
+            qstat.bytesReturned = 0
+            qstat.rowsReturned = 0
+            queriesToBeLogged.add(qstat)
+            queryWatcher.remove(queryId)
+        }
+        if (queriesToBeLogged.size > maxQueries) { run() }
+    }
+
+    override fun acceptQuery(
+        queryId: UUID,
+        user: User?,
+        headers: MutableMap<String, String>?,
+        apiVer: String?,
+        queryParams: Optional<MultivaluedMap<String, String>>?,
+        path: String?
+    ) {
+        var start: Long = System.currentTimeMillis()
+        var apiQuery: String? = constructAPIQuery(queryParams, path)
+        var userName: String? = user?.name
+        synchronized(queryWatcher) {
```

Comment:
You can use a ConcurrentHashMap and removed all the synchronized blocks that protect queryWatcher.